src/mjolnir/ferry_connections.cc

```diff
@@ -380,35 +380,46 @@ void ReclassifyFerryConnections(const std::string& ways_file,
           if (ret2.second) {
             outbound_path_found = true;
           }
+
+          // Reclassify the first/start edge if a connection to higher class roads
+          // is found. Do this AFTER finding shortest path so we do not immediately
+          // determine we hit the specified classification
+          sequence<Edge>::iterator element = edges[edge.second];
+          auto update_edge = *element;
+          if (ret1.second && ret2.second && update_edge.attributes.importance > kFerryUpClass) {
+            update_edge.attributes.importance = kFerryUpClass;
+            update_edge.attributes.reclass_ferry = remove_destonly;
+            element = update_edge;
+            total_count++;
+          }
         } else {
           // Check if oneway inbound to the ferry
           bool inbound =
               (edge.first.sourcenode_ == node_itr.position()) ? edge_rev_access : edge_fwd_access;
           auto ret = ShortestPath(node_itr.position(), end_node_idx, ways, way_nodes, edges, nodes,
                                   inbound, remove_destonly);
           total_count += ret.first;
-          if (ret.second > 0) {
+          if (ret.second) {
             if (inbound) {
               inbound_path_found = true;
             } else {
               outbound_path_found = true;
             }
           }
-        }
-        ferry_endpoint_count++;
 
-        // Reclassify the first/start edge if a connection to higher class roads
-        // is found. Do this AFTER finding shortest path so we do not immediately
-        // determine we hit the specified classification
-        sequence<Edge>::iterator element = edges[edge.second];
-        auto update_edge = *element;
-        if (inbound_path_found && outbound_path_found &&
-            update_edge.attributes.importance > kFerryUpClass) {
-          update_edge.attributes.importance = kFerryUpClass;
-          update_edge.attributes.reclass_ferry = remove_destonly;
-          element = update_edge;
-          total_count++;
+          // Reclassify the first/start edge if a connection to higher class roads
+          // is found. Do this AFTER finding shortest path so we do not immediately
+          // determine we hit the specified classification
+          sequence<Edge>::iterator element = edges[edge.second];
+          auto update_edge = *element;
+          if (ret.second && update_edge.attributes.importance > kFerryUpClass) {
+            update_edge.attributes.importance = kFerryUpClass;
+            update_edge.attributes.reclass_ferry = remove_destonly;
+            element = update_edge;
+            total_count++;
+          }
         }
+        ferry_endpoint_count++;
       }
 
       // Log cases where reclassification fails
```

Comment:
do we need to update the logging to not report false positives for failures when there are ferry connections that can only happen in or outbound? i dont know of any such mapped places but maybe there are ferries that only let you get off and drive away from the ferry and spots where you can only board the ferry (like separate entrance and exit). i suppose at the moment the code would need another else if (and test case) to cover the outbound only case?

Reply:
maybe we should just always run whichever ones have access and reclass whichever ones succeed. like if its forward access check outbound if it has reverse access check inbound and whichever ones work we'll reclass them. i havent looked at this in a long while so im probably oversimplifying it :smile:

Reply:
> i suppose at the moment the code would need another else if (and test case) to cover the outbound only case?

Actually this PR fixes such case as well, I'll cover it by test.

> maybe we should just always run whichever ones have access and reclass whichever ones succeed. like if its forward access check outbound if it has reverse access check inbound and whichever ones work we'll reclass them.

Well, currently it should work as the following:
- two way edge - reclassify it only if we found both inbound and outbound shortest paths via it
- one way edge - reclassify it if we found the corresponding (inbound for inbound edge and outbound for outbou nd edge) shortest paths

Reply:
> Actually this PR fixes such case as well, I'll cover it by test.

Please take a look - 078ac205b274beb3d4ac671a22e85d6c92f9301d